File: golang/02.go
package main

import (
	"strconv"
)

func Max(x int, y int) int {
    if x < y {
        return y
    }
    return x
}

func reverse(s string) string {
    rns := []rune(s) // convert to rune
    for i, j := 0, len(rns)-1; i < j; i, j = i+1, j-1 {
  
        // swap the letters of the string,
        // like first with last and so on.
        rns[i], rns[j] = rns[j], rns[i]
    }
  
    // return the reversed string.
    return string(rns)
}

func AddStrings(num1 string, num2 string) string {
	result := ""
	max := Max(len(num1), len(num2))
	diff := -1
	new_num1 := ""
	new_num2 := ""

	if len(num1) < max {
		diff = max - len(num1)	
		for i:=0; i<diff; i++ {
			new_num1 += "0"
		}
		new_num1 += num1
		num1 = new_num1
	}
	if len(num2) < max {
		diff = max - len(num2)	
		for i:=0; i<diff; i++ {
			new_num2 += "0"
		}
		new_num2 += num2
		num2 = new_num2
	}

	num1 = reverse(num1)
	num2 = reverse(num2)

	bring_forward := 0
	for index := 0; index < max; index++ {
		int1, err1 := strconv.Atoi(num1[index:index+1])
		int2, err2 := strconv.Atoi(num2[index:index+1])
		if err1 == nil && err2 == nil {
			sum := int1 + int2 + bring_forward
			if sum > 9 {
				result += strconv.Itoa(sum%10)
				bring_forward = int(sum/10)
			} else {
				result += strconv.Itoa(sum)
			}
		}
	}

	return reverse(result)
}

func main() {
	num1 := "11"
	num2 := "123"
	if AddStrings(num1, num2) != "134" {
		panic("Wrong answer for num1='11', num2='123'")
	}

	num1 = "456"
	num2 = "77"
	if AddStrings(num1, num2) != "533" {
		panic("Wrong answer for num1='456', num2='77'")
	}

	num1 = "0"
	num2 = "0"
	if AddStrings(num1, num2) != "0" {
		panic("Wrong answer for num1='0', num2='0'")
	}
}

File: python/02.py
# Time Complexity: O(max(N,M)). Given N is the number of elements in num1 and M is the number of elements in num2
# Space Complexity: O(1)
def AddStrings(num1, num2):
    _max = max(len(num1), len(num2))
    
    if len(num1) < _max:
        diff = _max-len(num1)
        new_num1 = "0"*diff + num1
        num1 = new_num1
    if len(num2) < _max:
        diff = _max-len(num2)
        new_num2 = "0"*diff + num2
        num2 = new_num2

    num1 = num1[::-1]
    num2 = num2[::-1]

    bring_forward = 0
    result = ""
    for index in range(_max): # O(max(N, M))
        _sum = int(num1[index]) + int(num2[index]) + bring_forward
        if _sum > 9:
            result += str(_sum%10)
            bring_forward = _sum//10
        else:
            result += str(_sum)

    return result[::-1]


num1 = "11"
num2 = "123"
AddStrings(num1, num2)
assert AddStrings(num1, num2) == "134", "Wrong answer for num1='11', num2='123'"

num1 = "456"
num2 = "77"
AddStrings(num1, num2)
assert AddStrings(num1, num2) == "533", "Wrong answer for num1='456', num2='77'"

num1 = "0"
num2 = "0"
AddStrings(num1, num2)
assert AddStrings(num1, num2) == "0", "Wrong answer for num1='0', num2='0'"